habitat-hitl/habitat_hitl/environment/controllers/llm_controller.py

```diff
@@ -0,0 +1,167 @@
+        self.initialize_environment_interface()
+        self.initialize_planner()
+        self.info: Dict[str, Any] = {}
+
+    def initialize_planner(self):
+        # NOTE: using instantiate here, but given this is planning for a single agent
+        # always will this ever be an option of Centralized vs Decentralized? Maybe
+        # DAG...?
+        # NOTE: assuming use of DecentralizedLLMPlanner here
+        planner = instantiate(self._config.evaluation.agents.agent_0.planner)
+        self.planner = planner(env_interface=self.environment_interface)
+        self.planner.agents = self.initialize_agents(
+            self._config.evaluation.agents
+        )
+
+    def initialize_agents(self, agent_configs):
+        agents = []
+        for _, agent_conf in agent_configs.items():
+            # Instantiate the agent
+            agent = Agent(
+                agent_conf.uid, agent_conf.config, self.environment_interface
+            )
+
+            # Make sure that its unique by adding to the set
+            agents.append(agent)
+        return agents
+
+    def initialize_environment_interface(self):
+        self.environment_interface = EnvironmentInterface(
+            self._config, gym_habitat_env=self._gym_habitat_env
+        )
+
+    def on_environment_reset(self):
+        # NOTE: the following ONLY resets self._test_recurrent_hidden_states,
+        # self._prev_actions and self._not_done_masks
+        # super().on_environment_reset()
+        self.environment_interface.reset_environment()
+        self.planner.reset()
+        if self._thread is not None:
+            self._thread.join()
+            self._thread = None  # noqa: F841
+            self._low_level_actions = {}
+
+        self.current_instruction = (
+            self.environment_interface.hab_env.current_episode.instruction
+        )
+        print(f"Instruction: {self.current_instruction}")
+        self._analysis_logger.debug(
+            f"------\nInstruction: {self.current_instruction}"
+        )
+        self._analysis_logger.debug("NEW EPISODE STARTING")
+        self._iter = 0
+
+    def _act(self, observations, *args, **kwargs):
+        # NOTE: update the world state to reflect the new observations
+        self.environment_interface.update_world_state(observations)
+        # NOTE: this is where the LLM magic happens, the agent is given the observations
+        # and it returns the actions for the agent
+        # TODO: looping needed here until a physical low-level-action is returned
+        # low_level_actions: Union[dict, np.ndarray] = {}
+        (
+            self._low_level_actions,
+            planner_info,
+            task_done,
+        ) = self.planner.get_next_action(
+            self.current_instruction,
+            observations,
+            self.environment_interface.world_graph,
+            verbose=True,
+        )
+        if task_done:
+            print("Task Done")
```

Comment:
I suppose that this would be the point where we hook up to HITL so that the agent can signal that its part of the task is done?

Would this also be called after the timeout? (`max_episode_steps`)

Reply:
The two conditions this would be called on:
1. If # of planning steps > max planning steps threshold
2. LLM reasons that the task is solved/can not be solved further